### PR TITLE
examples: maturin and setuptools_rust examples

### LINKED ISSUE #807: maturin examples
Now our examples are built by [setuptools-rust](https://github.com/PyO3/setuptools-rust), but since currently [maturin](https://github.com/PyO3/maturin) is (maybe more) widely used, we should also provide maturin examples.

### PR DESCRIPTION
**Pull request by davidhewitt**
This builds upon my experience writing the maturin and setuptools_rust examples in #1269 

This time I rearrange the example modules as follows:

```
pyo3_benchmarks -> pyo3-benchmarks
rustapi_module  -> pyo3-pytests
*new example*      maturin-starter
*new example*      setuptools-rust-starter
```

The two existing packages got a bit of a clean-up. The two new examples `maturin-starter` and `setuptools-rust-starter` are copies of each other tailored to the two different build systems.

I made good use of the trick in https://github.com/PyO3/pyo3/issues/1517#issuecomment-808664021 to make importing the submodules easy, and I wanted to show this off in all the examples. (e.g. `from maturin_starter.submodule import SubmoduleClass`).

Closes #807

### CODE REVIEW

**Comment by davidhewitt** [comment]
cc @konstin I've finally _almost_ completed a working set of maturin examples for PyO3!

Unfortunately it appears that `maturin` doesn't build under MSRV. That's ok as long as there's a wheel to install from. I'll need https://github.com/PyO3/maturin/pull/482 merged and included in a release, and then fingers crossed we'll be able to merge this PR here!

**Comment by davidhewitt** [comment]
Many thanks @konstin - CI is all green here using `maturin 0.10.0`! 🚀

File: examples/maturin-starter/pyproject.toml
Comment (konstin) [inline examples/maturin-starter/pyproject.toml:2]:
```suggestion
requires = ["maturin>=0.10,<0.11"]
```

File: examples/pyo3-pytests/pyproject.toml
Comment (konstin) [inline examples/pyo3-pytests/pyproject.toml:2]:
```suggestion
requires = ["maturin>=0.10,<0.11"]
```

File: examples/pyo3-pytests/tox.ini
Comment (konstin) [inline examples/pyo3-pytests/tox.ini:9]: 21.0.1 is released, does this work with pip from pypi?

**Comment by kngwyu** [comment]
Thanks!
Though I didn't fully check the code, I like the renaming, especially rustapi-module. Maybe we can close #1515 after merging this.

File: examples/setuptools-rust-starter/tox.ini
Comment (konstin) [inline examples/setuptools-rust-starter/tox.ini:9]: Same as above

**Review by konstin — APPROVED** [review]
Two small things, looks really nice otherwise

File: examples/pyo3-pytests/tox.ini
Comment (davidhewitt) [reply examples/pyo3-pytests/tox.ini:9]: Outdated comment sadly - the branch missed the merge window so will be released in `21.1`.

Comment (davidhewitt) [reply examples/pyo3-pytests/tox.ini:9]:
```suggestion
    # Use pip master with in-tree-build feature (to be released in pip 21.1)
```

File: examples/setuptools-rust-starter/tox.ini
Comment (davidhewitt) [reply examples/setuptools-rust-starter/tox.ini:9]:
```suggestion
    # Use pip master with in-tree-build feature (to be released in pip 21.1)
```

File: examples/maturin-starter/README.md
Comment (kngwyu) [inline examples/maturin-starter/README.md:3]: Let's add a link:
```suggestion
An example of a basic Python extension module built using PyO3 and [`maturin`](https://github.com/PyO3/maturin).
```

File: examples/pyo3-pytests/build.rs
Comment (kngwyu) [inline examples/pyo3-pytests/build.rs:27]: Why do we need this?

File: examples/setuptools-rust-starter/README.md
Comment (kngwyu) [inline examples/setuptools-rust-starter/README.md:3]:
```suggestion
An example of a basic Python extension module built using PyO3 and [`setuptools_rust`](https://github.com/PyO3/setuptools-rust).
```

**Review by kngwyu — APPROVED** [review]
Thanks!

File: examples/pyo3-pytests/build.rs
Comment (davidhewitt) [reply examples/pyo3-pytests/build.rs:27]: This used to be [in setup.py](https://github.com/PyO3/pyo3/blob/1751a7fb19b47a7dcc1f74d1ed442d74e4f39dbc/examples/rustapi_module/setup.py#L8), but we now build the tests with maturin instead of setuptools-rust.

We need it because some tests only compile on certain python / PyPy combinations.

IMO this `build.rs` method is better anyway because it would still work if we changed the examples back to `setuptools_rust`.

It's a little ugly to have to call a Python interpreter here. I have an idea how to split some of pyo3's `build.rs` logic into a separate crate `pyo3-build-config,` which this build.rs could then re-use. I've been torn for a while on whether that refactoring is worth the additional complexity; maybe I should hurry up and push that PR for a while so that others can see what they think.

**Review by messense — APPROVED** [review]

File: examples/pyo3-pytests/build.rs
Comment (davidhewitt) [reply examples/pyo3-pytests/build.rs:27]: I'm going to move this into a new issue so that it doesn't block this PR with all the new examples, which I think has value even if there's this this wart.

**Pull request merged**
